skimage/morphology/_rolling_ball_cy.pyx

```diff
@@ -0,0 +1,220 @@
+                     DTYPE_FLOAT[::1] kernel not None,
+                     DTYPE_FLOAT[::1] ellipsoid_intensity not None,
+                     Py_ssize_t[::1] img_shape not None,
+                     Py_ssize_t[::1] padded_img_shape not None,
+                     Py_ssize_t[::1] kernel_shape not None,
+                     Py_ssize_t num_threads=0):
+    """
+    apply_kernel_nan(img, kernel, ellipsoid_intensity, img_shape,
+                     padded_img_shape, kernel_shape, num_threads)
+
+    Apply a ND kernel to an ND image.
+
+    This function is the critical piece of code for 
+    `morphology.rolling_ellipsoid`. It was moved to cython for speed.
+
+    Parameters
+    ----------
+    img : (I) ndarray
+        A flat view into a padded image, e.g., from ``numpy.ravel()``.
+    kernel : (K) ndarray
+        A flat view into the kernel, e.g., from ``numpy.ravel()``. Indicates
+        if a pixel lies inside the ellipsoid. `kernel[x,y] == 1` if the pixel
+        is inside, ``kernel[x,y] == np.Inf`` otherwise.
+    ellipsoid_intensity : (K) ndarray
+        A flat view into the intensity of the ellipsoid, e.g., from 
+        ``numpy.ravel()``. Indicates the height/intensity of the ellipsoid at
+        position ``(x,y)``.
+    img_shape : (N) ndarray
+        The shape of the unflattened, unpadded image
+    padded_img_shape : (N) ndarray
+        The shape of the unflattened, padded image
+    kernel_shape : (N) ndarray
+        The shape of the unflattened kernel
+
+    Returns
+    -------
+    out_data : ndarray
+        the estimated background intensity. ``out_data.shape = img_shape``.
+
+    See Also
+    --------
+    rolling_ellipsoid
+    """
+    
+    cdef DTYPE_FLOAT[::1] out_data = np.zeros(np.prod(img_shape.base),
+                                              dtype=img.base.dtype)
+    cdef Py_ssize_t offset, offset_idx, out_data_size, ker_idx, img_idx
+    cdef DTYPE_FLOAT min_value, tmp
+    out_data_size = out_data.size
+
+    cdef Py_ssize_t ndim = kernel_shape.shape[0]
+    cdef Py_ssize_t kernel_outer = np.prod(kernel_shape[0:(ndim - 1)])
+    cdef Py_ssize_t kernel_inner = kernel_shape[ndim - 1]
+    cdef Py_ssize_t ker_idx_outer, ker_idx_inner
+
+    for offset_idx in prange(
+            out_data_size,
+            num_threads=num_threads,
+            nogil=True):
+        offset = ind2ind(offset_idx, 0, img_shape, padded_img_shape)
+        min_value = INFINITY
+
+        for ker_idx_outer in range(kernel_outer):
+            ker_idx_outer = ker_idx_outer * kernel_inner
+            img_idx = ind2ind(ker_idx_outer, offset, kernel_shape, 
+                              padded_img_shape)
+            for ker_idx_inner in range(kernel_inner):
+                ker_idx = ker_idx_outer + ker_idx_inner
+                tmp = (img[img_idx+ker_idx_inner] + ellipsoid_intensity[ker_idx]) * kernel[ker_idx]
+                if min_value > tmp or isnan(tmp):
+                    min_value = tmp
```

Comment:
@FirefoxMetzger if we are just going to propagate the nans, why is a separate function needed? Does the generated C arithmetic not comply with IEEE 754?

Reply:
Because we aren't using BLAS to compute the minimum over the kernel (it's significantly faster to not use BLAS and let the compiler optimize this kernel). This means that `min_value > nan == False` and we have to handle `nan` explicitly. This used to be significantly slower than not handling them because the compiler didn't vectorize. It's more similar now, but there is still a penalty for dealing with `nan`.  We can also break the loop in case of a `nan`, which brings the timings closer together (especially if there are a lot of `nan`s), but that optimization slows down the general case of being nan-free.

 [Sidenote: On windows compilation uses default flags, I'd assume the same holds for linux/unix, which is AVX. We could get a bit more performance on benchmarks overall if we'd allow compilation with AVX2/AVX512 or similar instruction sets if present.]

Benchmark:

```
[ 20.00%] ··· benchmark_morphology.RollingBall.time_rollingball                                                      ok
[ 20.00%] ··· ======== =========
               radius
              -------- ---------
                 25     220±0ms
                 50     450±0ms
                 75     925±0ms
                100     1.58±0s
                150     3.25±0s
                200     5.48±0s
              ======== =========

[ 30.00%] ··· benchmark_morphology.RollingBall.time_rollingball_nan                                                  ok
[ 30.00%] ··· ======== =========
               radius
              -------- ---------
                 25     194±0ms
                 50     405±0ms
                 75     950±0ms
                100     1.31±0s
                150     2.84±0s
                200     5.02±0s
              ======== =========
```